book-content/chapters/05-unions-literals-and-narrowing.md

```diff
@@ -0,0 +1,1856 @@
+
+#### Solution 2: Narrowing a Discriminated Union with a Switch Statement
+
+The first step is to clear out the `calculateArea` function and add the `switch` keyword and specify `shape.kind` as our switch condition:
+
+```typescript
+function calculateArea(shape: Shape) {
+  switch (shape.kind) {
+    case "circle": {
+      return Math.PI * shape.radius * shape.radius;
+    }
+    case "square": {
+      return shape.sideLength * shape.sideLength;
+    }
+    // Potential additional cases for more shapes
+  }
+}
+```
+
+As a nice bonus, TypeScript offers us autocomplete on the cases for the `switch` statement. This is a great way to ensure that we're handling all of the cases for our discriminated union.
+
+##### Not Accounting for All Cases
+
+As an experiment, comment out the case where the `kind` is `square`:
+
+```typescript
+function calculateArea(shape: Shape) {
+  switch (shape.kind) {
+    case "circle": {
+      return Math.PI * shape.radius * shape.radius;
+    }
+    // case "square": {
+    //   return shape.sideLength * shape.sideLength;
+    // }
+    // Potential additional cases for more shapes
+  }
+}
+```
+
+Now when we hover over the function, we see that the return type is `number | undefined`. This is because TypeScript is smart enough to know that if we don't return a value for the `square` case, the output will be `undefined` for any `square` shape.
+
+```typescript
+// hovering over `calculateArea` shows
+function calculateArea(shape: Shape): number | undefined;
+```
+
+Switch statements work great with discriminated unions!
+
+#### Solution 3: Destructuring a Discriminated Union of Tuples
+
+When you're done, your `APIResponse` type should look like this:
+
+```typescript
+type APIResponse = ["error", string] | ["success", User[]];
+```
+
+We've created two possible combinations for the `APIResponse` type. An error state, and a success state. And instead of objects, we've used tuples.
+
+You might be thinking - where's the discriminant? It's the first element of the tuple. This is what's called a discriminated tuple.
+
+And with this update to the `APIResponse` type, the errors have gone away!
+
+##### Understanding Tuple Relationships
+
+Inside of the `exampleFunc` function, we use array destructuring to pull out the `status` and `value` from the `APIResponse` tuple:
+
+```typescript
+const [status, value] = await fetchData();
+```
+
+Even though the `status` and `value` variables are separate, TypeScript keeps track of the relationships behind them. If `status` is checked and is equal to `"success"`, TypeScript can narrow down `value` to be of the `User[]` type automatically:
+
+```typescript
+// hovering over `status` shows
+const status: "error" | "success";
+```
+
+Note that this intelligent behavior is specific to discriminated tuples, and won't work with discriminated objects - as we saw in our previous exercise.
```

Comment:
Bollocks :P it didn't work only because properties were not shared among union members and then spread didn't work because spreading lost track of the relationship. This very example here would work fine with objects and destructuring - this behavior is, by no means, unique to tuples:
[TS playground](https://www.typescriptlang.org/play/?#code/C4TwDgpgBAqgzhATlAvFA3gKClAlgEwC4oA7AVwFsAjJAbm1IEMKJi5hFcSBzegX3qZQkKAEEACgEkAShDhgA9iQSoGAHwwMcOdo2Bk4xAERwyAYzNy4R+tu0A3RgBsyrWAkQBtALq3tfdU07HWA9A2MkRAVEGy0HZ1c2Di5eOIFMTEY4EBIzKAAzMlzgXCUCiGAzAAsAET1GAAoASmJxKIpcBAAeCRk5RWUIAD4gqA4QUZwzJXYoRH6Z6DRGAHdGXGByyqqGoyrgYDBDAHpjxjBcADoIAA9mMCcIS+mKY-x6oybBO1x8qAaAITzeSLS4KADWTUm2nm+kQJGhdl0+kMUCMkWiRgANHE7KcoJIKIxuFxoHAFCw8BQwFF7BB8FAMcgqowSPgnCkoFUkBBcfEXG4jKIEUyoAoLGREPN8Ni+ek7AE4tNlJt3qFUFBVutNsCBghLgArckkZrfGEVSUI9BQZHhNGmCxWbFQRwC4hqxhQeVeqBmPTVf5MqFYOywy0YG2hFERKWYrEuhKC4WM2PIcVmSXSoxevwBRVZHJ5QrFUoiu7Ux4AMSKZmao2Vs2ttrg8ddrh9yzWGy21TqoVNGRwv3+zdQKDQJnMljg1mDcTbED8OHxAD0APwMPiMpwqEP81xLqCrjc4POYIA)

Reply:
Based on that u might want to revisit paragraphs related to this behavior